packages/vx-legend/src/legends/Threshold.tsx

```diff
@@ -0,0 +1,85 @@
+import React from 'react';
+import Legend, { LegendProps } from './Legend';
+import { BaseInput, BaseOutput, ScaleThreshold, LabelFormatterFactory } from '../types';
+
+const formatZero = (label: unknown) => (label === 0 ? '0' : label || '');
+
+export type LegendThresholdProps<Datum extends BaseInput, Output extends BaseOutput> = {
+  labelDelimiter?: string;
+  labelLower?: string;
+  labelUpper?: string;
+  labelTransform?: LabelFormatterFactory<Datum, Output, ScaleThreshold<Datum, Output>>;
+  scale: ScaleThreshold<Datum, Output>;
+} & LegendProps<Datum, Output>;
+
+/** Default transform implicitly assumes that Datum is of type number. */
+function defaultTransform<Datum extends BaseInput, Output extends BaseOutput>({
+  labelDelimiter,
+  labelLower,
+  labelUpper,
+}: Pick<
+  LegendThresholdProps<Datum, Output>,
+  'labelDelimiter' | 'labelLower' | 'labelUpper'
+>): LabelFormatterFactory<Datum, Output, ScaleThreshold<Datum, Output>> {
+  return ({ scale, labelFormat }) => (d, i) => {
+    let [x0, x1] = scale.invertExtent(scale.range()[i]);

```

Comment:
this is messy, the problem is that the previous implementation used `scale.range()` and passed _that_ as the `domain` prop. that is no longer valid with types, and it's impossible to compute the right thresholds directly from the domain value `d`.

the domain of a `threshold` scale also is not guaranteed to be a `number` so we have extra checks for that here.

Reply:
after going through the d3 docs more, if `domain` has `n` items then the `range` of a valid scale should have `n+1`, so I think this is safe. I updated the logic to check that the item corresponding to index `i` is within the range, and added a comment. we can update in the future if need be.